### ISSUE TITLE
Rename asserters classes for PHP7

### ISSUE THREAD

**Issue by jubianchi** [atoum/atoum#452]
Since PHP7 integrates [scalar type hints](https://wiki.php.net/rfc/scalar_type_hints), some more reserved keywords were added:

> No new reserved words are added. The names int, integer, float, string, bool and boolean are recognised and allowed as type hints, and prohibited from use as class/interface/trait names (including with use and class_alias).

This affects atoum code on:

* asserters:
  * [integer](https://github.com/atoum/atoum/blob/master/classes/asserters/integer.php)
  * [float](https://github.com/atoum/atoum/blob/master/classes/asserters/float.php)
  * [string](https://github.com/atoum/atoum/blob/master/classes/asserters/string.php)
  * [boolean](https://github.com/atoum/atoum/blob/master/classes/asserters/boolean.php)
* mocks:
  * https://github.com/atoum/atoum/blob/master/classes/mock/generator.php#L806
  
Asserters should then be renamed and aliased to their old names ans the mock generator should be fixed:
* [x] `integer` => `phpInteger`
* [x] `float` => `phpFloat`
* [x] `string` => `phpString`
* [ ] `boolean` => `phpBoolean`
* [ ] The mock generator method should be fixed to include new reserved words.

If the [Context Sensitive Lexer](https://wiki.php.net/rfc/context_sensitive_lexer) RFC gets accepted, this issue will probably be closable.

**Comment by jubianchi** [comment]
After testingPHP 7's behavior, it seems that the RFC is not correct on which words are reserved (at least, reserved today).
I built a snippet which avels some code to generate classes/interfaces/traits/methods on the fly and see if a fatal is thrown:

```php
<?php

switch ($argv[1]) {
    case 'class':
    case 'interface':
    case 'trait':
        register_shutdown_function(function() {
            $error = error_get_last();

            if ($error !== null) {
                echo '   => ' . trim($error['message']) . PHP_EOL . PHP_EOL;
            }
        });

        $className = $argv[2];
        $check = $argv[1] . '_exists';
        $code = $argv[1] . ' ' . $className . ' {}';

        if ($argv[1] === 'class') {
            $code .= ' $f = new ' . $className . ';';
        }

        echo '## ' . $className . PHP_EOL;
        echo '   ```' . $code . '```' . PHP_EOL;

        @eval($code);

        if ($check($className)) {
            echo '   => OK' . PHP_EOL . PHP_EOL;
            exit(0);
        } else {
            echo '   => KO' . PHP_EOL . PHP_EOL;
        }

        exit(1);

    case 'method':
        register_shutdown_function(function() {
            $error = error_get_last();

            if ($error !== null) {
                echo '   => ' . trim($error['message']) . PHP_EOL . PHP_EOL;
            }
        });

        $methodName = $argv[2];
        $check = $argv[1] . '_exists';
        $code = 'class foo { public function ' . $methodName . '() {} } $f = new foo; $f->' . $methodName . '();';

        echo '## ' . $methodName . PHP_EOL;
        echo '   ```' . $code . '```' . PHP_EOL;

        @eval($code);

        if ($check('foo', $methodName)) {
            echo '   => OK' . PHP_EOL . PHP_EOL;
            exit(0);
        } else {
            echo '   => KO' . PHP_EOL . PHP_EOL;
        }

        exit(1);
}

if ($argv[2]) {
    register_shutdown_function(function() {
        $error = error_get_last();

        if ($error !== null) {
            echo '   => ' . trim($error['message']) . PHP_EOL . PHP_EOL;
        }
    });

    $methodName = $argv[2];
    $check = $argv[1] . '_exists';
    $code = $argv[1] . ' ' . $methodName . ' {}';

    echo '## ' . $methodName . PHP_EOL;
    echo '   ```' . $code . '```' . PHP_EOL;

    @eval($code);

    if ($check($methodName)) {
        echo '   => OK' . PHP_EOL . PHP_EOL;
    } else {
        echo '   => KO' . PHP_EOL . PHP_EOL;
    }

    exit(0);
}


$classNames = ['boolean', 'integer', 'string', 'float', 'bool', 'int'];

echo '# Classes:' . PHP_EOL;

foreach ($classNames as $methodName) {
    passthru('php ' . __FILE__ . ' class ' . $methodName);
}

echo '# Interfaces:' . PHP_EOL;

foreach ($classNames as $methodName) {
    passthru('php ' . __FILE__ . ' interface ' . $methodName);
}

echo '# Traits:' . PHP_EOL;

foreach ($classNames as $methodName) {
    passthru('php ' . __FILE__ . ' trait ' . $methodName);
}

echo '# Methods:' . PHP_EOL;

foreach ($classNames as $methodName) {
    passthru('php ' . __FILE__ . ' method ' . $methodName);
}

```

On `PHP 7.0.0-dev (cli) (built: Apr  7 2015 05:09:25) / Zend Engine v3.0.0-dev` the result is:

```
# Classes:
## boolean
   ```class boolean {} $f = new boolean;```
   => OK

## integer
   ```class integer {} $f = new integer;```
   => OK

## string
   ```class string {} $f = new string;```
   => Cannot use 'string' as class name as it is reserved

## float
   ```class float {} $f = new float;```
   => Cannot use 'float' as class name as it is reserved

## bool
   ```class bool {} $f = new bool;```
   => Cannot use 'bool' as class name as it is reserved

## int
   ```class int {} $f = new int;```
   => Cannot use 'int' as class name as it is reserved

# Interfaces:
## boolean
   ```interface boolean {}```
   => OK

## integer
   ```interface integer {}```
   => OK

## string
   ```interface string {}```
   => Cannot use 'string' as class name as it is reserved

## float
   ```interface float {}```
   => Cannot use 'float' as class name as it is reserved

## bool
   ```interface bool {}```
   => Cannot use 'bool' as class name as it is reserved

## int
   ```interface int {}```
   => Cannot use 'int' as class name as it is reserved

# Traits:
## boolean
   ```trait boolean {}```
   => OK

## integer
   ```trait integer {}```
   => OK

## string
   ```trait string {}```
   => Cannot use 'string' as class name as it is reserved

## float
   ```trait float {}```
   => Cannot use 'float' as class name as it is reserved

## bool
   ```trait bool {}```
   => Cannot use 'bool' as class name as it is reserved

## int
   ```trait int {}```
   => Cannot use 'int' as class name as it is reserved

# Methods:
## boolean
   ```class foo { public function boolean() {} } $f = new foo; $f->boolean();```
   => OK

## integer
   ```class foo { public function integer() {} } $f = new foo; $f->integer();```
   => OK

## string
   ```class foo { public function string() {} } $f = new foo; $f->string();```
   => OK

## float
   ```class foo { public function float() {} } $f = new foo; $f->float();```
   => OK

## bool
   ```class foo { public function bool() {} } $f = new foo; $f->bool();```
   => OK

## int
   ```class foo { public function int() {} } $f = new foo; $f->int();```
   => OK

```

**Comment by jubianchi** [comment]
My question is: do you think I should rename all asserters targeted by the RFC or just edit those that are impacted **today** ?

**Comment by jubianchi** [comment]
nobody ?

**Comment by agallou** [comment]
when using an asserter, I think it's easier to read ->string than ->phpString (same for the other cases).
But we can't keep those with php7.

It would be more consistant to have all asserters starting with phpSomething

So, IMHO i think a only asserters impacted today should be renamed. The other ones (https://github.com/jubianchi/atoum/tree/reserved-words-php7/classes/asserters) will stay easer to read, and everyone will understand why those asserters have php before (just like array today).

**Comment by Hywan** [comment]
I guess we have to ask internals if https://wiki.php.net/rfc/context_sensitive_lexer is going to be implemented or not.

**Comment by jubianchi** [comment]
@agallou this won't introduce any BC break thanks to atoum aliaser : only class will be renamed but asserters will still be usable with their actual names (ie `->string`, `->integer`, ...)
@Hywan given the voting state, the RFC passed.

**Comment by Hywan** [comment]
@jubianchi Without no implemention, so possibly implemented in PHP7.1. That's my fear.

**Comment by jubianchi** [comment]
@Hywan this is not a problem for us ;) To support PHP 7.0 we'll just have to rename asserter classes and create the required aliases... This has already been done in the associated PR which is already green on PHP 7 ;)

**Comment by mageekguy** [comment]
Hello all !
Firstly, awesome job, @jubianchi! i think you SHOULD put your script in a gist and post it on the IRC channel of internals@ (EFnet > #php.pecl).
Secondly, you SHOULD discuss on the IRC channel about that, to have insurance that you are on the right way.
My 2 cents.

**Comment by jubianchi** [comment]
After talking with people on EFnet > #php.pecl here are some facts:

* I used/linked the wrong RFC. The good one is at: https://wiki.php.net/rfc/scalar_type_hints_v5
* This version of the RFC is correct on the reserved keywords:
> No new reserved words are added. The names int, float, string and bool are recognised and allowed as type declarations, and prohibited from use as class/interface/trait names (including with use and class_alias).
* For now, no need to rename the `integer` and `boolean` asserters

I'm going to revert part of my patch, especially the renaming of the `integer` asserter.

**Comment by jubianchi** [comment]
closed by #454
